Move on to Mapnik 3.x

jlnr opened the pull request:
I'm trying to render OSM tiles from Java on Windows and OS X. I couldn't get mapnik-jni to link on Windows at first (same error as in issue #3), and when I did, things crashed as soon as I zoomed to an area that actually contained anything.

So I've adapted mapnik-jni to work with Mapnik 3; specifically, Mapnik 3.0.1 from Homebrew on OS X, and the 64-bit Mapnik 3.0.2 SDK for Windows linked to from this comment: https://github.com/mapnik/mapnik/issues/3014#issuecomment-127180550
(I can send a separate Pull Request for my MSVC 2015 project if this PR looks okay to you.)

All tests pass and everything works fine for me, but so far I've only loaded maps from XML and rendered them to PNG, so it is entirely possible that I've broken every other class on my way.
I'll comment on the commits in this PR about which I feel a little uneasy. Also, I haven't added any `#ifdef` magic to keep this compatible with Mapnik 2.x.

Commit:
Replace JNI lib loading logic w/ plain loadLibrary

Commit:
Use flags from config instead of JavaVM.framework

Commit:
Fix TestLoadLibrary depending on order of @Tests

Commit:
Updated auto-generated header (not sure why diff.)

Commit:
Conservatively update Java language level to 1.6

Commit:
Add missing reinterpret_cast<>

Commit:
Adapt library to work with the C++ interface of Mapnik 3.x

jlnr commented on build.xml:
```diff
@@ -5,9 +5,9 @@
 	<target name="compile">
 		<mkdir dir="build/classes"/>
 		<mkdir dir="csrc"/>
-		<javac srcdir="src" destdir="build/classes" debug="on" source="1.5" target="1.5" includeantruntime="false"/>
```
I've only actually tested this with Oracle's Java 8 VM on both OS X and Windows, which complained about `source="1.5"` (warnings). Maybe it would be better to update this to 1.8 or drop the `source` parameter altogether? I am not an Ant or JVM expert.

jlnr commented on csrc/Makefile.osx:
```diff
@@ -9,7 +9,8 @@ libmapnik-jni.jnilib: $(SOURCE_DEPENDS)
 		$(JAVA_CFLAGS) \
 		mapnikjni.cpp \
 		$(LDFLAGS) \
-		$(MAPNIK_LIBS) -framework JavaVM
```
The `JavaVM.framework` contains symlinks to a non-existent JDK 1.6 on my Mac; I'm not sure if that is the default state of Macs nowadays, or if I broke things by installing and uninstalling JDKs.

jlnr commented on csrc/class_datasource.cpp:
```diff
@@ -32,7 +32,30 @@ JNIEXPORT jobject JNICALL Java_mapnik_Datasource_getParameters
 		jstring key=env->NewStringUTF(iter->first.c_str());
-		boost::apply_visitor(translate_parameter_visitor(env, paramobject, key), iter->second);
+		translate_parameter_visitor visitor(env, paramobject, key);
+		// TODO - The call to visit() does not compile on MSVC 2015 (error C2783).
```
It doesn't compile on OS X, either. I'm not sure if I am even supposed to call `mapnik::value_type::visit` directly? It worked fine in another place.

jlnr commented on csrc/class_featureset.cpp:
```diff
@@ -57,10 +57,12 @@ JNIEXPORT jobjectArray JNICALL Java_mapnik_FeatureSet__1loadGeometries
-	unsigned count=(*fp)->num_geometries();
+	// Mapnik 3.x only holds one geometry<> per feature_impl.
+	// (In Mapnik 2.x this used to be a variable-length vector of geometries.)
+	unsigned count=1;
```
If I'm not misunderstanding things here, the JNI interface should be refactored to match the new C++ interface – but I didn't want to expand the scope of this PR for something that I'm not really sure about (I only want to render maps 😅).

jlnr commented on csrc/class_geometry.cpp:
```diff
@@ -8,11 +8,27 @@ JNIEXPORT jint JNICALL Java_mapnik_Geometry_getType
 }
 
+
+struct VertexCollector
```
What I'm trying to do here is to collect all the vertices in an arbitrary `geometry<double>` into an `std::vector`, using `geometry::vertex_processor`. I _think_ it should work, but it is certainly inefficient to do this on every call to `getNumPoints` or `getVertex`. => The (internal) Java interface should be refactored to do this more efficiently.

jlnr commented on src/mapnik/Mapnik.java:
```diff
@@ -18,37 +16,8 @@
 	private static boolean initialized;
 	private static boolean registered;
-	private static String loadedLibrary;
-	private static List<String> librarySearchNames=new ArrayList<String>();
```
I couldn't get mapnik-jni to find the JNI library on my Mac (even with Mapnik 2), so I've replaced the `librarySearchNames` logic by a single call to `loadLibrary` below. Now I point `-Djava.library.path=...` to the right directory and everything works.
I can take this commit out of the PR if this breaks mapnik-jni in other environments.

jlnr commented on test/mapnik/TestLoadLibrary.java:
```diff
@@ -13,18 +13,9 @@
  */
 public class TestLoadLibrary {
 	@Test
-	public void testStaticSettings() {
```
This test doesn't do anything now, so I've removed it. It also kept failing for me, I suspect that the test below was being run before this one, leaving Mapnik uninitialised.

pmauduit commented:
Hi, I'm very interested in this PR because I'm willing to generate a debian package (which provides a mapnik 3.x version of the library in the unstable version of the distribution), so that I can use it easily in a java project, any news so far ? May I help ?

FYI, trying to build, I got the following error:

```
     [exec] mapnikjni.cpp:11:31: fatal error: mapnik/graphics.hpp: No such file or directory
```

jlnr commented:
(EDIT: snipped misleading section; leaving the bottom half of this comment because it might still be a good idea)

I guess it would be cleaner _and_ fix your compiler bug to change the `csrc/Makefile.{osx,linux}` to use `mapnik-config --cflags` instead, in the same way that the Makefiles already use `mapnik-config --ldflags`. If you successfully patch `Makefile.linux` and let me know the diff, I'll push it into this PR too, in the hopes that it will eventually be merged.

pmauduit commented:
Strange, I don't have the file `graphics.hpp` in my include directory, using the 3.0.9 dev package:

``` bash
% dpkg -l | grep ii | grep mapnik 
ii  libmapnik-dev                          3.0.9+ds-1                             amd64        C++ toolkit for developing GIS applications (devel)
```

I'll dig in further ...

jlnr commented:
At a first glance, the v3.0.5 tag in mapnik/mapnik does not seem to include `graphics.hpp`. Strange. Let me know what you find, maybe I'm working with outdated files somewhere.

jlnr commented:
Hey @pmauduit. Sorry for being completely misleading in my earlier comment. There is no `mapnik/graphics.hpp` in Mapnik 3, but it did exist in Mapnik 2. You are probably compiling a checkout of this repository, not my fork. Compare `mapnikjni.cpp` here:

https://github.com/springmeyer/mapnik-jni/blob/master/csrc/mapnikjni.cpp#L11

with the file in my branch:

https://github.com/jlnr/mapnik-jni/blob/mapnik-3/csrc/mapnikjni.cpp#L11

Hope that helps.

pmauduit commented:
I was pretty sure to compile against your current branch, but I guess it deserves a triple-check :-) I will dig in further ASAP.

pmauduit commented:
I actually was on the good branch, but with some probable autogenerated files from previous trials onto the master one ... After everything cleaned up, I was able to compile.

There is a failing test though revealed by `make test`, but haven't been further yet:

```
    [junit] Test mapnik.TestLayer FAILED
```

I did a sample Java app that generates my hometown's highways as png, and it works like a charm ; now struggling with the specific debian packaging stuff :)

jlnr commented:
If I'm not mistaken then one of the tests required me starting a PostGIS server in order to pass. That might be it.

Commit:
Fix 64-bit pointer truncation on Windows

The sweet irony of an ASSERT() helper causing so much trouble :)

`long` is 32-bit even on 64-bit Windows:
https://msdn.microsoft.com/en-us/library/s3f49ktz.aspx

Andy511823558 commented:
How do I get the 'mapnik-jni.dll'? Thanks!
